src/hotspot/share/services/memMapPrinter.cpp

```diff
@@ -0,0 +1,303 @@
+  if (!MemTracker::enabled()) {
+    return false;
+  }
+  NMTRegionSearchWalker walker(from, to);
+  VirtualMemoryTracker::walk_virtual_memory(&walker);
+  if (walker._found.has_any()) {
+    // The address range we may be asked about may be the result of VMA folding:
+    // Two adjacent memory mappings that happen to have the same property will be folded
+    // by the kernel into a single VMA.
+    // Since that can seriously confuse readers of this mapping output, we try to find
+    // out if the mapping is used for multiple purposes. We mark those VMAs with "(*)"
+    // and print all NMT region markings.
+    for (int i = 0; i < mt_number_of_types; i++) {
+      const MEMFLAGS flag = (MEMFLAGS)i;
+      if (walker._found.has_flag(flag)) {
+        st->print("%s", get_shortname_for_nmt_flag(flag));
+        if (flag == mtThreadStack) {
+          print_thread_details_for_supposed_stack_address(from, to, st);
+        }
+        st->print(" ");
+      }
+    }
+    if (walker._match_type == NMTRegionSearchWalker::MatchType::vma_superset) {
+      st->print(" (*)");
+    }
+    return true;
+  }
+
+  return false;
+}
+
+static void print_legend(outputStream* st) {
+#define DO(flag, shortname, text) st->print_cr("%10s    %s", shortname, text);
+  NMTFLAGS_DO(DO)
+#undef DO
+}
+
+MappingPrintClosure::MappingPrintClosure(outputStream* st, bool human_readable, jlong timeout_at) :
+    _out(st), _human_readable(human_readable), _timeout_at(timeout_at), _total_count(0), _total_vsize(0) {}
+
+bool MappingPrintClosure::do_it(const MappingPrintInformation* info) {
+  _total_count++;
+  _out->print(PTR_FORMAT " - " PTR_FORMAT " ", p2i(info->from()), p2i(info->to()));
+  const size_t size = pointer_delta(info->to(), info->from(), 1);
+  _total_vsize += size;
+  if (_human_readable) {
+    _out->print(PROPERFMT " ", PROPERFMTARGS(size));
+  } else {
+    _out->print("%11zu", size);
+  }
+  assert(info->from() <= info->to(), "Invalid VMA");
+  _out->fill_to(53);
+  info->print_OS_specific_details_heading(_out);
+  _out->fill_to(70);
+  ask_nmt_about(info->from(), info->to(), _out);
+  _out->fill_to(100);
+  info->print_OS_specific_details_trailing(_out);
+  _out->cr();
+
+  return _timeout_at > os::javaTimeNanos(); // false if timeout
+}
+
+void MemMapPrinter::print_all_mappings(outputStream* st, bool human_readable) {
+
+  st->print_cr("Memory mappings:");
+  if (!MemTracker::enabled()) {
+    st->print_cr(" (For full functionality, please enable Native Memory Tracking)");
+  }
+  st->cr();
+  print_legend(st);
+  st->print_cr("(*) - Mapping contains data from multiple regions");
+  st->cr();
+  pd_print_header(st);
+  // Under rare circumstances the process memory map may be insanely large and/or fragmented. We cap
+  // the absolute runtime of printing to blocking other VM operations too long.
+  const jlong timeout_at = os::javaTimeNanos() +
+                           ((jlong)(SafepointTimeoutDelay * NANOSECS_PER_MILLISEC) / 2);
```

Comment:
Is AbortVMOnSafepointTimeoutDelay/2 timeout useful for real world cases?

I assume the client who would need this feature, would need it for a reason and that means memory hungry processes?

Instead of blocking VM and having to use the timeout, can we:

A. take a snapshot of NMT virtual memory region and process it without blocking the VM

B. divide this process into smaller chunks, each of which will be guaranteed to finish before AbortVMOnSafepointTimeoutDelay triggers

C. temporarily increase AbortVMOnSafepointTimeoutDelay and revert it back afterwords?

In case A,B we would get back a very close, but not exact view of memory, but wouldn't this be better than getting back precise, but cut off view?

Case C would give us what the users wants assuming they are OK with waiting a bit longer?

And adding malloc'ed memory here would be make this issue much worse.

Reply:
The whole abort-on-timeout thing I added because I run the command in a VMOp, and it runs in a VMOp because I want to print information about GC threads (so, not because of NMT) and I was not sure how long the GC Thread* objects are guaranteed to live. Otherwise, I would not have bothered. AbortVMOnSafepointTimeoutDelay is by default 10 seconds, that is really plenty.

But.... 

Gaaah! :-( I just tried to dump a million mappings in a test and found that the limiting factor is the output stream size of the jcmd bufferedStream. It looks like jcmd buffers the whole $%!& output before sending it over the wire. For hypothetical situations with millions of mappings, this would mean an additional memory footprint in the GB range. Thats quite annoying.

I opened https://bugs.openjdk.org/browse/JDK-8319055 to track that, but have not high hopes of someone actually having the time for it. A potential solution using temporary buffers (an outputStream variant that under the hood buffers its content in a temp file) would be simple, but a bit embarrassing and may need convincing folks.

I need to think about this...